### PR TITLE
Feat(eos_cli_config_gen): Option to set RCF for route bgp ipv4/ipv6 mulit-cast and VRF

### PR DESCRIPTION
**Pull request by Vibhu-gslab**
## Change Summary

Option to set RCF for route bgp ipv4/ipv6 mulit-cast and VRF.

## Related Issue(s)

part of https://github.com/aristanetworks/avd/issues/2397

## Component(s) name

`arista.avd.eos_cli_config_gen`

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md [hunk 1]
```diff
@@ -1338,6 +1346,9 @@
         redistribute_routes:
           - source_protocol: <str; required; unique>
             route_map: <str>
+
+            # RCF function name without parenthesis.
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md]: can you confirm again that rcf is supported for address_family_ipv4_multicast->redistribute_routes as am not able to get in device
```
wan1-dc1(config-router-bgp)#address-family ipv4 multicast 
wan1-dc1(config-router-bgp-af)#redistribute connected ?
  route-map  Name a route map
  <cr>       

wan1-dc1(config-router-bgp-af)#redistribute connected 
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md [hunk 2]
```diff
@@ -1429,6 +1440,12 @@
             # IPv6 prefix "A:B:C:D:E:F:G:H/I"
           - prefix: <str; required; unique>
             route_map: <str>
+        redistribute_routes:
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1610]: same as above

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1610]: RCF is only available for "isis" here:

```
s1-spine1(config-router-bgp)#address-family ipv6 multicast 
s1-spine1(config-router-bgp-af)#redistribute isis rcf ?
  FUNCNAME()  Routing control function name
  ```
  I have added a condition in j2 file:
  https://github.com/aristanetworks/avd/pull/3821/files#diff-b389418c101262db3e5545c20387bfabb1eda7c9ad8a7a1bc21b212be2252b93R987
  
  May be we can add this in description, I will do that.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md (hunk 1 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md]: RCF is only available for "isis" here:

```
s1-spine1(config-router-bgp)#address-family ipv4 multicast 
s1-spine1(config-router-bgp-af)#redistribute isis rcf ?
  FUNCNAME()  Routing control function name
  ```

I have added a condition in j2 file:
https://github.com/aristanetworks/avd/pull/3821/files#diff-b389418c101262db3e5545c20387bfabb1eda7c9ad8a7a1bc21b212be2252b93R837

May be we can add this in description, I will do that.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/router-bgp-evpn-vpn-import-pruning.md [hunk 3]
```diff
@@ -117,7 +117,7 @@ ASN Notation: asplain
 
 | VRF | Route-Distinguisher | Redistribute |
 | --- | ------------------- | ------------ |
-| TENANT_A_PROJECT01 | 192.168.255.3:11 | connected |
+| TENANT_A_PROJECT01 | 192.168.255.3:11 | connected<br>static<br>isis |
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/router-bgp-evpn-vpn-import-pruning.md:120]: can we have it comma separated

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md [hunk 4]
```diff
@@ -1623,12 +1644,20 @@
                 route_targets:
                   - <str>
                 route_map: <str>
+
+                # RCF function name with parenthesis.
+                # Example: MyFunction(myarg).
+                rcf: <str>
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1854]: There is additional RCF support for vpn-ipv4/ipv6 address families 
```
wan1-dc1(config-router-bgp-vrf-Mahesh)#route-target import vpn-ipv4 rcf FUN1() ?
  vpn-route  RCF filter VPN routes
  <cr>       

wan1-dc1(config-router-bgp-vrf-Mahesh)#route-target import vpn-ipv4 rcf FUN1() vpn-route ?
  filter-rcf  RCF filter routes

wan1-dc1(config-router-bgp-vrf-Mahesh)#route-target import vpn-ipv4 rcf FUN1() vpn-route filter-rcf FUN2() ?
  <cr>  
```
Also RCF not supported for flow-spec-vpn-pv4/ipv6 address families

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 [hunk 5]
```diff
@@ -834,6 +834,10 @@ router bgp {{ router_bgp.as }}
 {%                 set redistribute_route_cli = "redistribute " ~ redistribute_route.source_protocol %}
 {%                 if redistribute_route.route_map is arista.avd.defined %}
 {%                     set redistribute_route_cli = redistribute_route_cli ~ " route-map " ~ redistribute_route.route_map %}
+{%                 elif redistribute_route.source_protocol == "isis" %}
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: For isis redistribute_route default it takes level-2 however we have the option to set on different level. Can you check in your EOS version that it has level visibility or not
```
router bgp 65000
   address-family ipv4 multicast
      redistribute isis level-2 route-map ROUTMAP
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 [hunk 6]
```diff
@@ -972,6 +976,22 @@ router bgp {{ router_bgp.as }}
 {%             endif %}
       {{ network_cli }}
 {%         endfor %}
+{%         for redistribute_route in router_bgp.address_family_ipv6_multicast.redistribute_routes | arista.avd.natural_sort('source_protocol') %}
+{%             if redistribute_route.source_protocol is arista.avd.defined %}
+{%                 set redistribute_route_cli = "redistribute " ~ redistribute_route.source_protocol %}
+{%                 if redistribute_route.source_protocol == "bgp" %}
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: bgp is not for ipv6 multicast address family. Even if we do it reflect to global config

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/router-bgp-evpn-vpn-import-pruning.md (hunk 3 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/router-bgp-evpn-vpn-import-pruning.md:120]: "\<br\>" works better. As discussed.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 (hunk 5 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: Please don't bloat this PR. It is about RCF, so just implement what we do elsewhere for redistribution.

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: Won't be addressing in this PR

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md (hunk 4 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1854]: Added!

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 (hunk 6 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2]: Corrected it.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg [hunk 7]
```diff
@@ -70,11 +70,15 @@ router bgp 65101
    vrf TENANT_A_PROJECT01
       rd 192.168.255.3:11
       route-target import evpn 11:11
+      route-target import evpn rcf RT_IMPORT_AF_RCF()
       route-target export evpn 11:11
+      route-target export evpn rcf RT_EXPORT_AF_RCF()
       router-id 192.168.255.3
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg:76]: To get the same order as EOS we have to iterate on same loop twice or may be append in different list then combine it. Below is the output from EOS
```
vrf TENANT_A_PROJECT01
      route-target import evpn 11:11
      route-target export evpn 11:11
      route-target import evpn rcf RT_IMPORT_AF_RCF()
      route-target export evpn rcf RT_EXPORT_AF_RCF()
      redistribute connected
      redistribute static rcf Router_BGP_Static()
```

Comment (MaheshGSLAB) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg:76]: cc: @ClausHolbechArista

**Review by Shivani-gslab — APPROVED** [review]
Other than the ordering @MaheshGSLAB has mentioned, LGTM!

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 [hunk 8]
```diff
@@ -1272,8 +1300,17 @@ router bgp {{ router_bgp.as }}
...
-{%                 if address_family.route_map is arista.avd.defined %}
+{%                 if address_family.address_family in ['evpn', 'vpn-ipv4', 'vpn-ipv6'] %}
+{%                     if address_family.rcf is arista.avd.defined %}
+{%                         if address_family.vpn_route_filter_rcf is arista.avd.defined and address_family.address_family in ['vpn-ipv4', 'vpn-ipv6'] %}
+      route-target export {{ address_family.address_family }} rcf {{ address_family.rcf }} vpn-route filter-rcf {{ address_family.vpn_route_filter_rcf }}
+{%                         else %}
+      route-target export {{ address_family.address_family }} rcf {{ address_family.rcf }}
+{%                         endif %}
+{%                     endif %}
+{%                     if address_family.route_map is arista.avd.defined %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:1369]: Please check if both route-map and rcf can be on EOS, and if not, let route-map "win" by check that first and then have an `elif` for rcf.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:1369]: same for import

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg (hunk 7 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg:76]: please check the ordering for route-map. If both are done afterwards, then adjust the template to loop twice. If not, leave it as-is.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/router-bgp-evpn-vpn-import-pruning.cfg:76]: ah no, we cannot in the minor release, so leave it, and we will have to revisit in 5.0 work.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2 (hunk 8 above)
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:1369]: Both can be on EOS :)
```
router bgp 1
   vrf 1
      route-target import vpn-ipv4 rcf RT_IMPORT_AF_RCF()
      route-target export vpn-ipv4 rcf RT_EXPORT_AF_RCF()
      route-target import vpn-ipv4 route-map RT_IMPORT_AF_RM
      route-target export vpn-ipv4 route-map RT_EXPORT_AF_RM
```

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:1369]: all good.

**Review by ClausHolbechArista — APPROVED** [review]

**Pull request merged**